R/html_dependencies.R

```diff
@@ -194,16 +194,26 @@ html_reference_path <- function(path, lib_dir, output_dir) {
 # return the html dependencies as an HTML string suitable for inclusion
 # in the head of a document
 html_dependencies_as_string <- function(dependencies, lib_dir, output_dir) {
+  # Only copy deps with path
+  dependencies_href <- filter_dependencies(dependencies, "href")
+  dependencies <- filter_dependencies(dependencies, "file")
 
   if (!is.null(lib_dir)) {
     dependencies <- lapply(dependencies, copyDependencyToDir, lib_dir)
     dependencies <- lapply(dependencies, makeDependencyRelative, output_dir)
   }
-  return(renderDependencies(dependencies, "file", encodeFunc = identity,
+  #return(
+  tags <- renderDependencies(dependencies, "file", encodeFunc = identity,
                             hrefFilter = function(path) {
                               html_reference_path(path, lib_dir, output_dir)
                             })
-  )
+  #)
+  HTML(paste0(c(tags, renderDependencies(dependencies_href, "href")), collapse = "\n"))
+}
+
+filter_dependencies <- function(dependencies, src_type = "file") {
+  deps <- lapply(dependencies, function(dep) if(!is.null(dep$src[[src_type]])) return(dep))
```

Comment:
consider changing this to accept a vector of src_types, like:

```r
function(dep) if (any(!vapply(dep$src[src_types], is.null, logical(1)))) return(dep)
```

I'd probably write this lambda as multiple lines for readability:
```r
function(dep) {
  undefined_types <- vapply(dep$src[src_types], is.null, logical(1))
  if ( ! all(undefined_types) ) return(dep)
}
```

The above should circumvent reordering dependencies.

Reply:
We can also not filter and process dependencies in a for-loop as `renderDependencies` is doing internally. 
This would allow to apply the correct `renderDependencies` function, because only difference we need to make currently between href and files are about `encodeFunc=` and `hrefFilter=` in `renderDependencies()`. 
We could also revisit that to simplify maybe. 

This is considering my comment above regarding href deps being ignored (https://github.com/rstudio/rmarkdown/pull/1948#pullrequestreview-628985594) which confirmed by @jcheng5 in https://github.com/rstudio/rmarkdown/pull/1805#discussion_r608191632

Reply:
Something like this based on your change. 
````diff
diff --git a/R/html_dependencies.R b/R/html_dependencies.R
index 9f92dacb..6a81d9ee 100644
--- a/R/html_dependencies.R
+++ b/R/html_dependencies.R
@@ -194,26 +194,27 @@ html_reference_path <- function(path, lib_dir, output_dir) {
 # return the html dependencies as an HTML string suitable for inclusion
 # in the head of a document
 html_dependencies_as_string <- function(dependencies, lib_dir, output_dir) {
-  # Only copy deps with path
-  dependencies_href <- filter_dependencies(dependencies, "href")
-  dependencies <- filter_dependencies(dependencies, "file")

   if (!is.null(lib_dir)) {
-    dependencies <- lapply(dependencies, copyDependencyToDir, lib_dir)
-    dependencies <- lapply(dependencies, makeDependencyRelative, output_dir)
+    dependencies <- lapply(dependencies, copyDependencyToDir,
+                           output_dir = lib_dir, mustWork = FALSE)
+    dependencies <- lapply(dependencies, makeDependencyRelative,
+                           basepath = output_dir, mustWork = FALSE)
   }
-  #return(
-  tags <- renderDependencies(dependencies, "file", encodeFunc = identity,
-                            hrefFilter = function(path) {
-                              html_reference_path(path, lib_dir, output_dir)
-                            })
-  #)
-  HTML(paste0(c(tags, renderDependencies(dependencies_href, "href")), collapse = "\n"))
-}
-
-filter_dependencies <- function(dependencies, src_type = "file") {
-  deps <- lapply(dependencies, function(dep) if(!is.null(dep$src[[src_type]])) return(dep))
-  deps[!sapply(deps, is.null)]
+  html <- c()
+  for (dep in dependencies) {
+    tags <-
+      if (is.null(dep$src[["file"]])) {
+        renderDependencies(list(dep), "href")
+      } else {
+        renderDependencies(list(dep), "file", encodeFunc = identity,
+                           hrefFilter = function(path) {
+                             html_reference_path(path, lib_dir, output_dir)
+                           })
+      }
+    html <- c(html, tags)
+  }
+  HTML(paste(c(html, collapse = "\n")))
 }
````

I am still wondering if we really need to differentiate `renderDependencies` call. 🤔 
it is a pretty old code (since **rmarkdown** 0.6 maybe) so hard to tell.

Reply:
I kept the differentiation for now and went with the code change above